pa11ycrawler/reporter.py

```diff
@@ -0,0 +1,260 @@
+# -*- coding: utf-8 -*-
+"""
+This module defines the ways in which the data obtained from the
+Pa11ySpider and Pa11yPipleline is processed.
+"""
+import json
+import os
+import shutil
+
+from abc import ABCMeta, abstractmethod
+from mako.lookup import TemplateLookup
+
+
+class ReporterBaseClass(object):
+    """
+    Defines reporting directory structure and settings.
+    """
+    def __init__(self, settings):
+        self.settings = settings
+        self.reports_dir = settings.get('PA11YCRAWLER_REPORTS_DIR')
+        self.keep_existing = settings.get(
+            'PA11YCRAWLER_KEEP_EXISTING_REPORTS')
+
+        self.source_dir = os.path.join(self.reports_dir, 'source')
+        self.results_dir = os.path.join(self.reports_dir, 'results')
+        self.html_dir = os.path.join(self.reports_dir, 'html')
+        self.results_map_file = os.path.join(self.results_dir, 'pa11y-summary.json')
+        self.results_map = {}
+
+
+class Pa11yAbstractSpiderReporter():
+    """
+    Defines the expected methods. These are the points during a crawl when
+    the reported will be accessed.
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def add_results(self, item):
+        """
+        Called for every `A11yItem` processed through `Pa11yPipeline`.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def pre_process_reports(self):
+        """
+        Called before Pa11ySpider starts crawling.
+        Use for setting up and cleaning report directories, etc.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def post_process_reports(self):
+        """
+        Called when Pa11ySpider has finished crawling.
+        All items that will be processed will be done at this point.
+        Use for aggregating stats accross pages, etc.
+        """
+        raise NotImplementedError
+
+
+class Pa11yGenericReporter(ReporterBaseClass, Pa11yAbstractSpiderReporter):
+    """
+    Defines the default way of processing the Pa11y results.
+    """
+    def _save_html(self, item):
+        """
+        Saves the source html to `self.html_dir` if `PA11YCRAWLER_SAVE_SOURCE`
+         is `True`.
+        """
+        if self.settings.get('PA11YCRAWLER_SAVE_SOURCE'):
+            if not os.path.exists(self.source_dir):
+                os.makedirs(self.source_dir)
+
+            filepath = os.path.join(
+                self.source_dir, item['filename'] + '.html')
```

Comment:
I can.  I'm not sure how helpful it would be for readability since the filenames are already non-meaningful to a person (they're md5 hashes of the urls).  It will help ensure files aren't accidentally overwritten if using the `--pa11ycrawler-keep-existing-reports` flag though.